abi_encode.go

```diff
@@ -0,0 +1,377 @@
+		return nil, encodeErr(v, "string")
+	}
+	return packBytesSlice([]byte(v.String()), v.Len())
+}
+
+func packBytesSlice(buf []byte, l int) ([]byte, error) {
+	len, err := encodeNum(reflect.ValueOf(l))
+	if err != nil {
+		return nil, err
+	}
+	return append(len, rightPad(buf, (l+31)/32*32)...), nil
+}
+
+func packNum(offset int) []byte {
+	n, _ := encodeNum(reflect.ValueOf(offset))
+	return n
+}
+
+func encodeNum(v reflect.Value) ([]byte, error) {
+	switch v.Kind() {
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		return toU256(new(big.Int).SetUint64(v.Uint())), nil
+
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return toU256(big.NewInt(v.Int())), nil
+
+	case reflect.Ptr:
+		if v.Type() != bigIntT {
+			return nil, encodeErr(v.Elem(), "number")
+		}
+		return toU256(v.Interface().(*big.Int)), nil
+
+	case reflect.Float64:
+		return encodeNum(reflect.ValueOf(int64(v.Float())))
+
+	case reflect.String:
+		n, ok := new(big.Int).SetString(v.String(), 10)
+		if !ok {
+			n, ok = new(big.Int).SetString(v.String()[2:], 16)
+			if !ok {
+				return nil, encodeErr(v, "number")
+			}
+		}
+		return encodeNum(reflect.ValueOf(n))
+
+	default:
+		return nil, encodeErr(v, "number")
+	}
+}
+
+func encodeBool(v reflect.Value) ([]byte, error) {
+	if v.Kind() != reflect.Bool {
+		return nil, encodeErr(v, "bool")
+	}
+	if v.Bool() {
+		return leftPad(one.Bytes(), 32), nil
+	}
+	return leftPad(zero.Bytes(), 32), nil
+}
+
+func encodeErr(v reflect.Value, t string) error {
+	return fmt.Errorf("failed to encode %s as %s", v.Kind().String(), t)
+}
+
+// nolint
+func mapFromStruct(v reflect.Value) (reflect.Value, error) {
+	res := map[string]interface{}{}
+	typ := v.Type()
+	for i := 0; i < v.NumField(); i++ {
+		f := typ.Field(i)
+		if f.PkgPath != "" {
+			continue
+		}
+
+		tagValue := f.Tag.Get("abi")
+		if tagValue == "-" {
+			continue
+		}
+
+		name := strings.ToLower(f.Name)
+		if tagValue != "" {
+			name = tagValue
+		}
+		if _, ok := res[name]; !ok {
+			res[name] = v.Field(i).Interface()
+		}
+	}
+	return reflect.ValueOf(res), nil
+}
+
+var (
+	tt256   = new(big.Int).Lsh(big.NewInt(1), 256)   // 2 ** 256
+	tt256m1 = new(big.Int).Sub(tt256, big.NewInt(1)) // 2 ** 256 - 1
+)
+
+// U256 converts a big Int into a 256bit EVM number.
+func toU256(n *big.Int) []byte {
+	b := new(big.Int)
+	b = b.Set(n)
+
+	if b.Sign() < 0 || b.BitLen() > 256 {
+		b.And(b, tt256m1)
+	}
+
+	return leftPad(b.Bytes(), 32)
+}
+
+func padBytes(b []byte, size int, left bool) []byte {
+	l := len(b)
+	if l == size {
+		return b
+	}
+	if l > size {
+		return b[l-size:]
+	}
+	tmp := make([]byte, size)
+	if left {
+		copy(tmp[size-l:], b)
+	} else {
+		copy(tmp, b)
+	}
+	return tmp
+}
+
+// nolint
+func leftPad(b []byte, size int) []byte {
+	return padBytes(b, size, true)
+}
+
+func rightPad(b []byte, size int) []byte {
+	return padBytes(b, size, false)
+}
+
+func decodeHex(str string) ([]byte, error) {
```

Comment:
I think this name is too general to be package private in the hedera package. I suggest renaming it to a more specific name.

Reply:
Im open for suggestions.

Reply:
Maybe you can leave it as is and delete the decodeHex in the `ethereum_transaction` test(Assuming it will do the same). This way we won't have to change the test and no merge conflict will follow.

Reply:
The test function panics, while the actual implementation is returning an error. I made the test function an internal variable, this should fix the compilation error.

Reply:
The test util func is removed.